src/main/java/com/ge/predix/uaa/token/lib/FastTokenServices.java

```diff
@@ -121,164 +118,70 @@ private void initTokenKeysCache(final long ttlMillis) {
-                resourceIds.addAll(values);
-            }
-
-            Set<GrantedAuthority> clientAuthorities = new HashSet<>();
-            if (claims.containsKey("client_authorities")) {
-                @SuppressWarnings("unchecked")
-                Collection<String> values = (Collection<String>) claims.get("client_authorities");
-                clientAuthorities.addAll(getAuthorities(values));
-            }
-
-            BaseClientDetails clientDetails = new BaseClientDetails();
-            clientDetails.setClientId(remoteClientId);
-            clientDetails.setResourceIds(resourceIds);
-            clientDetails.setAuthorities(clientAuthorities);
-            clientAuthentication.setResourceIdsAndAuthoritiesFromClientDetails(clientDetails);
+            throw new InvalidBearerTokenException("Malformed Access Token", e);
+        } catch (JOSEException e) {
+            throw new RuntimeException(e);
         }
-
-        Map<String, String> requestParameters = new HashMap<>();
-        if (isStoreClaims()) {
-            for (Map.Entry<String, Object> entry : claims.entrySet()) {
-                if (entry.getValue() != null && entry.getValue() instanceof String) {
-                    requestParameters.put(entry.getKey(), (String) entry.getValue());
-                }
-            }
-        }
-
-        if (claims.containsKey(Claims.ADDITIONAL_AZ_ATTR)) {
-            try {
-                requestParameters.put(Claims.ADDITIONAL_AZ_ATTR,
-                        JsonUtils.writeValueAsString(claims.get(Claims.ADDITIONAL_AZ_ATTR)));
-            } catch (JsonUtils.JsonUtilException e) {
-                throw new IllegalStateException("Cannot convert access token to JSON", e);
-            }
+        LOG.debug("Access token is valid.");
+        AbstractAuthenticationToken token = this.jwtAuthenticationConverter.convert(jwtDecoder.decode(accessToken));
+        if (token != null && token.getDetails() == null) {
+            token.setDetails(authentication.getDetails());
+            token.setAuthenticated(true);
+        } else {
+            throw new InvalidBearerTokenException("Invalid Access Token");
         }
-        clientAuthentication.setRequestParameters(Collections.unmodifiableMap(requestParameters));
-
-        Authentication userAuthentication = getUserAuthentication(claims, scope);
-
-        clientAuthentication.setApproved(true);
-        return new OAuth2Authentication(clientAuthentication.createOAuth2Request(), userAuthentication);
+        LOG.debug("Authentication successful.");
+        return token;
     }
 
-    private void verifyIssuer(final String iss) {
+    protected void verifyIssuer(final String iss) {
         Assert.notEmpty(this.trustedIssuers, "Trusted issuers must be defined for authentication.");
 
         if (!this.trustedIssuers.contains(iss)) {
-            throw new IssuerNotTrustedException("The issuer '" + iss + "' is not trusted "
```

Comment:
cant we use more specific IssuerNotTrustedException?

Reply:
There is no exception in the latest version, also the message is clear to the end user, so I think we are good with this.